File: backend/models/email.py
"""Email models for FastAPI Email Helper API."""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class EmailBase(BaseModel):
    """Base email model with standardized field names.
    
    Field standards:
    - content: Email content/body text
    - received_time: When email was received
    - ai_category: AI-assigned category
    """
    subject: str
    sender: str
    recipient: Optional[str] = None
    content: Optional[str] = None
    received_time: Optional[datetime] = None


class EmailCreate(EmailBase):
    """Email creation model."""
    id: str  # Email ID from email provider


class EmailUpdate(BaseModel):
    """Email update model."""
    category: Optional[str] = None
    confidence: Optional[float] = Field(None, ge=0.0, le=1.0)


class EmailInDB(EmailBase):
    """Email model as stored in database."""
    id: str
    ai_category: Optional[str] = None  # Standardized: 'ai_category' (AI classification)
    category: Optional[str] = None  # User-corrected category (for accuracy tracking)
    confidence: Optional[float] = None
    processed_at: datetime
    user_id: Optional[int] = None

    model_config = {"from_attributes": True}


class Email(EmailBase):
    """Email model for API responses.
    
    Uses standardized field names:
    - content: Email body text
    - received_time: When email was received
    - ai_category: AI-assigned category
    - category: User-corrected category (for accuracy tracking)
    """
    id: str
    ai_category: Optional[str] = None  # AI classification
    category: Optional[str] = None  # User-corrected category
    confidence: Optional[float] = None
    processed_at: datetime

    model_config = {"from_attributes": True}


class EmailClassification(BaseModel):
    """Email classification result."""
    ai_category: str
    confidence: float = Field(ge=0.0, le=1.0)
    reasoning: Optional[str] = None
    action_items: List[str] = []
    priority: Optional[str] = None


class EmailBatch(BaseModel):
    """Batch email processing request."""
    emails: List[Dict[str, Any]]
    context: Optional[str] = None


class EmailBatchResult(BaseModel):
    """Batch email processing result."""
    processed_count: int
    successful_count: int
    failed_count: int
    results: List[EmailClassification]
    errors: List[str] = []

File: backend-go/internal/models/email.go
package models

import "time"

// Email represents an email message
type Email struct {
	ID               string    `json:"id"`
	Subject          string    `json:"subject"`
	Sender           string    `json:"sender"`
	Recipient        string    `json:"recipient,omitempty"`
	Body             string    `json:"body"`
	Content          string    `json:"content,omitempty"`
	ReceivedTime     time.Time `json:"received_time,omitempty"`
	Date             time.Time `json:"date"`
	IsRead           bool      `json:"is_read"`
	HasAttachments   bool      `json:"has_attachments"`
	Importance       string    `json:"importance"`
	Categories       []string  `json:"categories"`
	ConversationID   string    `json:"conversation_id,omitempty"`
	ConversationCount int      `json:"conversation_count,omitempty"`

	// AI Classification
	AICategory    string  `json:"ai_category,omitempty"`
	AIConfidence  float64 `json:"ai_confidence,omitempty"`
	AIReasoning   string  `json:"ai_reasoning,omitempty"`
	OneLineSummary string `json:"one_line_summary,omitempty"`

	// User correction
	Category string `json:"category,omitempty"`
}

// EmailListResponse represents paginated email list
type EmailListResponse struct {
	Emails  []Email `json:"emails"`
	Total   int     `json:"total"`
	Offset  int     `json:"offset"`
	Limit   int     `json:"limit"`
	HasMore bool    `json:"has_more"`
}

// EmailOperationResponse represents result of email operation
type EmailOperationResponse struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
	EmailID string `json:"email_id"`
}

// EmailFolderResponse represents list of email folders
type EmailFolderResponse struct {
	Folders []EmailFolder `json:"folders"`
	Total   int           `json:"total"`
}

// EmailFolder represents an email folder
type EmailFolder struct {
	Name          string `json:"name"`
	UnreadCount   int    `json:"unread_count"`
	TotalCount    int    `json:"total_count"`
	FolderPath    string `json:"folder_path,omitempty"`
}

// ConversationResponse represents an email conversation thread
type ConversationResponse struct {
	ConversationID string  `json:"conversation_id"`
	Emails         []Email `json:"emails"`
	Total          int     `json:"total"`
}

// UpdateClassificationRequest represents classification update
type UpdateClassificationRequest struct {
	Category         string `json:"category"`
	ApplyToOutlook   bool   `json:"apply_to_outlook"`
}

// CategoryFolderMapping maps categories to folders
type CategoryFolderMapping struct {
	Category      string `json:"category"`
	FolderName    string `json:"folder_name"`
	StaysInInbox  bool   `json:"stays_in_inbox"`
}

// BulkApplyRequest for bulk operations
type BulkApplyRequest struct {
	EmailIDs       []string `json:"email_ids"`
	ApplyToOutlook bool     `json:"apply_to_outlook"`
}

// BulkApplyResponse response from bulk operations
type BulkApplyResponse struct {
	Success    bool     `json:"success"`
	Processed  int      `json:"processed"`
	Successful int      `json:"successful"`
	Failed     int      `json:"failed"`
	Errors     []string `json:"errors,omitempty"`
}

// SyncEmailRequest for syncing emails to database
type SyncEmailRequest struct {
	Emails []Email `json:"emails"`
}

// EmailBatch for batch processing
type EmailBatch struct {
	Emails []interface{} `json:"emails"` // Can be Email objects or just IDs
}

// EmailBatchResult result of batch processing
type EmailBatchResult struct {
	ProcessedCount  int                      `json:"processed_count"`
	SuccessfulCount int                      `json:"successful_count"`
	FailedCount     int                      `json:"failed_count"`
	Results         []EmailProcessingResult  `json:"results"`
	Errors          []string                 `json:"errors,omitempty"`
}

// EmailProcessingResult individual email processing result
type EmailProcessingResult struct {
	EmailID     string        `json:"email_id"`
	Category    string        `json:"category"`
	Confidence  float64       `json:"confidence"`
	Reasoning   string        `json:"reasoning"`
	ActionItems []string      `json:"action_items"`
	Priority    string        `json:"priority"`
	Error       string        `json:"error,omitempty"`
}
